add go example recipe (#2547)
conda-forge/conda-forge.github.io: docs/maintainer/example_recipes/go.md

  name: ${{ name|lower }}
  version: ${{ version }}

source:
  url: https://github.com/example-package/${{ name }}/archive/refs/tags/v${{ version }}.tar.gz
  sha256: 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
  target_directory: src

build:
  number: 0
  script:
    - cd src
    - go-licenses save . --save_path ../library_licenses
    - if: unix
      then: go build -v -o $PREFIX/bin/saml-auth-proxy
      else: go build -v -o %LIBRARY_BIN%\saml-auth-proxy.exe

requirements:
  build:
    - ${{ compiler("go") }}
    - go-licenses

tests:
  - script: example-package -help
  - package_contents:
    bin:
      - example-package
    strict: true

about:
  homepage: https://github.com/example-package/example-package
  summary: Summary of the package.
  description: |
    Description of the package
  license: MIT
  license_file:
    - src/LICENSE
    - library_licenses/
  documentation: https://pkg.go.dev/github.com/example-package/example-package
  repository: https://github.com/example-package/example-package

extra:
  recipe-maintainers:
    - LandoCalrissian
```

This recipe template supports different features:

- Package licenses of statically linked libraries.

> - Ensure only binary is created by using `strict: true` in the `package_contents` tests.

Author: technically, this did not go through a CEP yet, right @wolfv?
Should we skip this for now?

Reviewer: yep, we haven't ratified this yet ...

Reviewer 2: It's important to note here that a significant number of Go packages also ship completions/man pages, so those will need to be allowed in the package as well, but I guess they can just be specified in the `package_contents` list as we well?

---
title: 'Go packages'
---

If you want to package a Go package to conda-forge, you can use this recipe template:

```yaml
context:
  name: example-package
  version: "0.1.0"

package:
  name: ${{ name|lower }}
  version: ${{ version }}

source:
  url: https://github.com/example-package/${{ name }}/archive/refs/tags/v${{ version }}.tar.gz
  sha256: 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
  target_directory: src

build:
  number: 0
  script:
    - cd src
    - go-licenses save . --save_path ../library_licenses
    - if: unix
      then: go build -v -o $PREFIX/bin/example-package

>       else: go build -v -o %LIBRARY_BIN%\example-package.exe

Author: added your comments to the bottom: https://github.com/conda-forge/conda-forge.github.io/pull/2547/commits/b81e5ccf061962dcacfbdfb049315317795d5ec7

requirements:
  build:
    - ${{ compiler("go") }}
    - go-licenses

tests:
  - script: example-package -help
  - package_contents:
    bin:
      - example-package
    strict: true

about:
  homepage: https://github.com/example-package/example-package
  summary: Summary of the package.
  description: |
    Description of the package
  license: MIT
  license_file:
    - src/LICENSE
    - library_licenses/
  documentation: https://pkg.go.dev/github.com/example-package/example-package
  repository: https://github.com/example-package/example-package

extra:
  recipe-maintainers:
    - LandoCalrissian
```

This recipe template supports different features:

- Package licenses of statically linked libraries.
- Ensure only binary is created by using `strict: true` in the `package_contents` tests.

> If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).

Author: are you sure? it sounds to me like the canonical way of doing things would be `compiler('cgo')` as it's in the variant files in staged-recipes (and then probably in conda-smithy as well?): https://github.com/conda-forge/staged-recipes/blob/db1ed99bbe4aab539b3b012164842e52f9f315d3/.ci_support/linux64.yaml#L15

> If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).

Author: if we adjust this, we need to adjust the `compiler('go')` in the top as well to a `compiler('go-nocgo')` IMO

> If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).

Reviewer 2: Yes, we should use `compiler('go-nocgo')` instead - it is more clear as to what variant is being used!

> If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).

Author: [`4850fda` (#2547)](https://github.com/conda-forge/conda-forge.github.io/pull/2547/commits/4850fdaca4fc761213187a0c36eeaf8e99e8e932)

  sha256: 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
  target_directory: src

build:
  number: 0
  script:
    - cd src
    - go-licenses save . --save_path ../library_licenses
    - if: unix
      then: go build -v -o $PREFIX/bin/example-package -ldflags="-s -w"
      else: go build -v -o %LIBRARY_BIN%\example-package.exe -ldflags="-s"

requirements:
  build:
    - ${{ compiler("go-nocgo") }}
    - go-licenses

tests:
  - script: example-package -help
  - package_contents:
    bin:
      - example-package
    strict: true

about:
  homepage: https://github.com/example-package/example-package
  summary: Summary of the package.
  description: |
    Description of the package
  license: MIT
  license_file:
    - src/LICENSE
    - library_licenses/
  documentation: https://pkg.go.dev/github.com/example-package/example-package
  repository: https://github.com/example-package/example-package

extra:
  recipe-maintainers:
    - LandoCalrissian
```

This recipe template supports different features:

- Package licenses of statically linked libraries.
- Ensure only binary is created by using `strict: true` in the `package_contents` tests.

> If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("go-cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).

Reviewer 2: ```suggestion
If your package requires `cgo` instead of `go-nocgo`, you can use `${{ compiler("go-cgo") }}` instead to build the package. By default, the `go-nocgo` compiler [is used](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64.yaml).  Using `${{ compiler("go-cgo")) }}` also requires `${{ compiler("c") }}` and `${{ stdlib("c") }}` and may require `${{ compiler("cxx") }}` if C++ code is compiled.  C/C++ build tools such as `make`, `autoconf`, `automake`, `libtool` or `cmake` may also be needed.
```

Sometimes, `go-licenses` might fail to detect licenses for some packages. In such cases, you can manually download the license file from the official source and add `--ignore github.com/bad-package/bad-package` to the `go-licenses` invokation. See [here](https://github.com/conda-forge/k9s-feedstock/blob/7929e0d86c829ba2ca172f08926f9fb7e6398247/recipe/recipe.yaml) for an example.

> Some packages ship multiple binaries. Then, the `go build` command needs to be run for each binary.

Reviewer 2: ```suggestion
Some packages ship multiple binaries, in which case the `go build` command needs to be run separately for each binary.
```

> It is increasingly common for the go build command to require specifying a subdirectory, so it would look something like:

Reviewer 2: ```suggestion
It is increasingly common for the go build command to require specifying a subdirectory, in which case the build steps would be slightly modified as shown below:
```

```bash
go-licenses save ./cmd/example-package --save-path ../library_licenses
go build -v -o ${PREFIX}/bin/example-package -ldflags="-s -w" ./cmd/example-package
```

> 
> This can vary from project to project.

Reviewer 2: ```suggestion
```

Some older Go packages that are not using Go modules yet can be converted by using:

```bash
go mod init
go mod tidy
```

> Maybe some additional work is required to make older packages buildable with Go modules.

Reviewer 2: ```suggestion
Additional steps maybe required depending on which Go dependency manager the project has been using.
```
